### ISSUE TITLE
listaddressesbyasset has Inconsistent boolean values vs documentation, should these be in JSON format?

### ISSUE THREAD

**Issue by ravenlandpush** [RavenProject/Ravencoin#535]
### Expected behaviour
raven-cli should return the number of assets via listaddressesbyasset "ASSET_NAME" false, but help documentation does not match actual usage.

### Actual behaviour
See the screenshot.

![inconsistent_boolean_documentation](https://user-images.githubusercontent.com/47545876/53781868-88da4080-3f02-11e9-84e6-162641de55b6.png)

## What version of bitcoin-core are you using?
ravencoin2.2.2.0, latest

### Machine specs:
- OS: linux

As title suggests. Just a correction really for what is in the help file differing from usage of bools.  Guessing this should be in some json format rather than as help file suggests. 

Cheers !

**Comment by ravenlandpush** [comment]
[ramdisk@bridge database2]$ raven-cli listaddressesbyasset "WWW.RAVENLAND.ORG" |wc -l
50002

This does not quite look right here. I would like to get the full listing again, I'm guessing that 50,000 is restricted limit purposefully to prevent a request taking too long to page?

It's good idea, but how to make a request for the next 50,000? The help documentation doesn't seem to match what output one would expect to achieve this. Can you give an example in JSon and/or update the help files with the correct boolean format for this listaddressesbyassetfunction?

Thanks and sorry to be a pain!

**Comment by Scotty0448** [comment]
listaddressesbyasset "CATE" false   works correctly for me, returning the
full list of 8033 addresses.  I can't see in your screenshot what could be
different though.  I'm using Linux ravend 2.2.2.0.  You need the new asset
index enabled (assetindex=1) for this command to work correctly, but I
believe an error message is presented if you don't.


Use the final "start" parameter to page through address lists that are
greater than 50,000. So to get the second page, set start to 50000, third
page 10000, etc.

**Comment by Scotty0448** [comment]
That should have been 100000, not 10000.

You'll also need to set the "count" parameter to be able to set the "start"
parameter.  So it can be less than 50000 if desired.  Just set "start" to
be in multiples of whatever you set "count" to be to get subsequent pages.

**Comment by ravenlandpush** [comment]
Hey scotty, thanks for replying mate. Appreciate it.

So, this was actually experienced after upgrading my asset explorer backend ravend to 2.2.2.0 and then having to do a -reindex-chainstate -assetindex start for ravend. I actually encountered this issue afterwards. 

I discover with the latest version it only outputs in batches of 50,000, which seems to be intended behaviour? 

See  

[ramdisk@bridge database2]$ raven-cli listaddressesbyasset "WWW.RAVENLAND.ORG" |wc -l
50002

I noticed the issue after upgrading when the sql database exports were light and specifically many balances were missing from asset WWW.RAVENLAND.ORG. Cate is find because it never reaches >50,000 and hit the limit. I can see that you can specify the limit using bool listaddressbyasset ASSET_NAME false int int but despite trying all variants I wasn't able to get it to work. Downgrading to the December version immediately fixed it. Although I'm aware that it's much slower having lacked the optimizations and so on in the cache. 

It's annoying because it seems to be just me. Maybe its something specific to CentOS. Once I finish an export using the old version for the backend I'll retry to reproduce this error since it is pretty weird. 

- push

**Comment by Scotty0448** [comment]
Yes, a limit of 50,000 at a time is the intended behavior now.  Before that
limit was put in place the listaddressesbyasset command for me was freezing
ravend on the WWW.RAVENLAND.ORG asset.  I waited a little over 10 minutes
once and it still didn't return and all functions of ravend were frozen the
whole time until I eventually killed it.  The assumption was that it was
somehow due to its large number of addresses so a limit was put in place.

By the way, the same limit and similar parameters were put in place for the
listassetbalancesbyaddress command.

**Comment by ravenlandpush** [comment]
Yep we had the same problems so I expected this to be the case. The rpc calls do return eventually, but what can happen if multiple commands become queued they get locked whilst the first in is waiting to return. We've had it before, despite contrary belief it is possible to get the command to function well in the older version and that is the way we had it albeit a bit slow due to the way memory is paged in it. Because of this apparent issue I was trying listaddressesbyasset bool, and it was rejecting both true, false 1 and 0 values, which is the real reason I open this issue. 

what I still can't explain is this 'Json value is not boolean as expected'. despite trying 'true' 'false' '1' and '0' values after CATE the syntax I was using was correct.

It important to note the examples given in raven-cli help listaddressesbyasset

raven-cli listaddressesbyasset "CATE" false 2 0
raven-cli listaddressesbyasset "CATE" false
All gave me this 'Json value is not boolean as expected'. Naturally i am running with -assetindex as you  say I should, because this function is not available without it. So that is what i am interested in, since I can't at present on the new version use it seemingly and had to downgrade . Maybe I miss something.. but I don't thinkso?

I thought it was just that the datatype for the RPC function wasnt matching, but if I understand you and msft right I'm the only one whose encountered this issue. It's bugging me so I am going to see if I get hte same issue on another box. I will replicate if I can.. it nearly 7am so might be a little while.

Cheers !
push

**Comment by ravenlandpush** [comment]
oddly enough I wasn't able to recreate it on the second env

[raven@ravenland-france .raven]$ raven-cli listaddressesbyasset CATE false | wc -l
8035
[raven@ravenland-france .raven]$ raven-cli listaddressesbyasset CATE true
8033

It was definitely broken on the first for one reason or another once the backend is finishes i'll see if its recreatable.

**Comment by ravenlandpush** [comment]
I can confirm I have been able to replicate this issue.


[ramdisk@bridge raven-2.2.2.0]$ cd bin
[ramdisk@bridge bin]$ ./ravend -assetindex

..
2019-03-05 07:55:50 dnsseed thread exit
2019-03-05 07:55:50 receive version message: /Ravencoin:2.2.2/: version 70020, b                                                                                                                                                             locks=606720, us=82.30.173.147:52182, peer=0
2019-03-05 07:55:56 receive version message: /Ravencoin:2.2.2(ravencore-sl)/: ve                                                                                                                                                             rsion 70020, blocks=606720, us=82.30.173.147:49666, peer=1
2019-03-05 07:55:58 UpdateTip: new best=000000000000252e54ed2704fc6e586677d226dc                                                                                                                                                             8ac069f8292378c5dcdfc177 height=606720 version=0x30000000 log2_work=66.142516 tx                                                                                                                                                             =3957952 date='2019-03-05 07:55:02' progress=0.999999 cache=0.0MiB(28txo)
^C
[ramdisk@bridge ~]$ raven-cli listaddressesbyasset CATE true
error code: -1
error message:
JSON value is not a boolean as expected
[ramdisk@bridge ~]$ raven-cli listaddressesbyasset CATE | head -n 2
{
  "R9HC7XkCwnQA5dQZ18BntXgUe9ESuALU3J": 100.00,
[ramdisk@bridge ~]$ raven-cli listaddressesbyasset CATE true
error code: -1
error message:
JSON value is not a boolean as expected
[ramdisk@bridge ~]$ raven-cli listaddressesbyasset CATE false
error code: -1
error message:
JSON value is not a boolean as expected
[ramdisk@bridge ~]$ raven-cli -V
error: too few parameters (need at least command)
[ramdisk@bridge ~]$ raven-cli --version
Raven Core RPC client version v2.2.0.0-0e0cf44
[ramdisk@bridge ~]$ ls -la
total 77192


At first., I thought that I had made a booboo and was using a mismatched cli version. 

[ramdisk@bridge bin]$ ./ravend --version
Raven Core Daemon version v2.2.0.0-0e0cf44
Copyright (C) 2009-2018 The Raven Core developers

Can confirm they match.

I wanted to be doubly doubly sure

[ramdisk@bridge bin]$ ./raven-cli listaddressesbyasset CATE true
error code: -1
error message:
JSON value is not a boolean as expected
[ramdisk@bridge bin]$ pwd
/home/ramdisk/raven-2.2.0.0/bin

What's going on do you think?

**Comment by ravenlandpush** [comment]
```2019-03-05 08:02:55 Raven version v2.2.2.0-49a5509
2019-03-05 08:02:55 InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2019-03-05 08:02:55 Validating signatures for all blocks.
2019-03-05 08:02:55 Setting nMinimumChainWork=0000000000000000000000000000000000000000000000000000000000000000
2019-03-05 08:02:55 Using the 'sse4' SHA256 implementation
2019-03-05 08:02:55 Default data directory /home/ramdisk/.raven
2019-03-05 08:02:55 Using data directory /home/ramdisk/.raven
2019-03-05 08:02:55 Using config file /home/ramdisk/.raven/raven.conf
2019-03-05 08:02:55 Using at most 125 automatic connections (1024 file descriptors available)
2019-03-05 08:02:55 Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2019-03-05 08:02:55 Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2019-03-05 08:02:55 Using 4 threads for script verification
2019-03-05 08:02:55 scheduler thread start

2019-03-05 08:05:06 HTTP: creating work queue of depth 16
2019-03-05 08:05:06 Config options rpcuser and rpcpassword will soon be deprecated. Locally-run instances may remove rpcuser to use cookie-based auth, or may be replaced with rpcauth. Please see share/rpcuser for rpcauth auth generation.
2019-03-05 08:05:06 HTTP: starting 4 worker threads
2019-03-05 08:05:06 init message: Verifying wallet(s)...
2019-03-05 08:05:06 Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2019-03-05 08:05:06 Using wallet wallet.dat
2019-03-05 08:05:06 CDBEnv::Open: LogDir=/home/ramdisk/.raven/database ErrorFile=/home/ramdisk/.raven/db.log
2019-03-05 08:05:06 scheduler thread start
2019-03-05 08:05:06 Block index database configuration:
2019-03-05 08:05:06 * Using 2 MB files
2019-03-05 08:05:06 Cache configuration:
2019-03-05 08:05:06 * Using 2.0MiB for block index database
2019-03-05 08:05:06 * Using 8.0MiB for chain state database
2019-03-05 08:05:06 * Using 440.0MiB for in-memory UTXO set (plus up to 286.1MiB of unused mempool space)
2019-03-05 08:05:06 init message: Loading block index...
2019-03-05 08:05:06 Opening LevelDB in /home/ramdisk/.raven/blocks/index
2019-03-05 08:05:06 Opened LevelDB successfully
2019-03-05 08:05:06 Using obfuscation key for /home/ramdisk/.raven/blocks/index: 0000000000000000
2019-03-05 08:05:06 Opening LevelDB in /home/ramdisk/.raven/assets
2019-03-05 08:05:06 Opened LevelDB successfully
2019-03-05 08:05:06 Using obfuscation key for /home/ramdisk/.raven/assets: 0000000000000000
2019-03-05 08:05:06 Loaded Assets from database without error
Cache of assets size: 1250
2019-03-05 08:05:40 LoadBlockIndexDB: last block file = 36
2019-03-05 08:05:40 LoadBlockIndexDB: last block file info: CBlockFileInfo(blocks=4762, size=103178564, heights=601009...606728, time=2019-03-01...2019-03-05)
2019-03-05 08:05:40 Checking all blk files are present...
2019-03-05 08:05:40 LoadBlockIndexDB: transaction index disabled
2019-03-05 08:05:40 LoadBlockIndexDB: asset index enabled
2019-03-05 08:05:40 LoadBlockIndexDB: address index disabled
2019-03-05 08:05:40 LoadBlockIndexDB: timestamp index disabled
2019-03-05 08:05:40 LoadBlockIndexDB: spent index disabled
2019-03-05 08:05:40 Opening LevelDB in /home/ramdisk/.raven/chainstate
2019-03-05 08:05:40 Opened LevelDB successfully
2019-03-05 08:05:40 Using obfuscation key for /home/ramdisk/.raven/chainstate: 9442a6540d861c03
2019-03-05 08:05:40 Loaded best chain: hashBestChain=0000000000002fd67fa01af3cad2447a42060dd5dfe6fa8bb1630258b5632fa3 height=606728 date=2019-03-05 08:04:39 progress=0.999998
2019-03-05 08:05:40 init message: Rewinding blocks...
2019-03-05 08:05:41 init message: Verifying blocks...
2019-03-05 08:05:41 Verifying last 6 blocks at level 3
2019-03-05 08:05:41 [0%]...[16%]...[33%]...[50%]...[66%]...[83%]...[99%]...[DONE].
2019-03-05 08:05:41 No coin database inconsistencies in last 7 blocks (102 transactions)
2019-03-05 08:05:41  block index           34926ms
2019-03-05 08:05:41 init message: Loading wallet...
2019-03-05 08:05:41 nFileVersion = 2020200
2019-03-05 08:05:41 Keys: 2001 plaintext, 0 encrypted, 2001 w/ metadata, 2001 total
2019-03-05 08:05:41  wallet                   34ms
2019-03-05 08:05:41 setKeyPool.size() = 2000
2019-03-05 08:05:41 mapWallet.size() = 0
2019-03-05 08:05:41 mapAddressBook.size() = 0
2019-03-05 08:05:41 mapBlockIndex.size() = 606736
2019-03-05 08:05:41 nBestHeight = 606728
2019-03-05 08:05:41 Connection Manager: StartBound to [::]:8767
2019-03-05 08:05:41 Bound to 0.0.0.0:8767

```

**Comment by ravenlandpush** [comment]
since it is odd and you might be wondering this now. I'm using CentOS 7.

```[ramdisk@bridge ~]$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC)
[ramdisk@bridge ~]$ uname -a
Linux bridge.localdomain 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by Scotty0448** [comment]
A few messages back you were checking if the versions match, which they do,
but they are version 2.2.0 rather than 2.2.2.

    [ramdisk@bridge ~]$ raven-cli --version
    Raven Core RPC client version v2.2.0.0-0e0cf44

    [ramdisk@bridge bin]$ ./ravend --version

    Raven Core Daemon version v2.2.0.0-0e0cf44
    Copyright (C) 2009-2018 The Raven Core developers

**Comment by blondfrogs** [comment]
@ravenlandpush During our testing, we tested with Mac, Windows, and Ubuntu. We didn't test with CentOS 7 which seems to be the issue for you in this case. 

**JSON value is not a boolean as expected** message is coming from when raven-cli tries to convert the value give to a boolean and it fails. 

I am running **Raven Core RPC client version v2.2.2.0**

Can you please verify your build version with ./raven-cli --version

**Comment by ravenlandpush** [comment]
I do apologise. Although I briefly mentioned I'd thought about it and shown that I checked the CLI version matched the running version of raven-d, if we look closely at my logs they do not match, because of the later changes, the raven-cli needs to be updated. I thought I had done this, and checked it but alas at 7am I wasn't exactly paying attention.

TLDR is that people should note that in later versions the raven-cli is expecting different input than ravend because of the changes made in the 2.2.0 and 2.2.2.0 versions. Kind of obvious really, had I not missed the output I would have arrested this yesterday.

You can close this out. A little oversight on my behalf here.  Thanks for the lightning fast assist blondfrogs, and well spotted scotty !

Cheers &
Best wishes,
Adam

**Comment by blondfrogs** [comment]
Closing as version 2.2.2 fixes this issue.
